sklearn/preprocessing/_encoders.py

```diff
@@ -52,103 +61,112 @@ def _check_X(self, X):
             needs_validation = True
 
-        n_samples, n_features = X.shape
+        n_features = X.shape[1]
         X_columns = []
 
         for i in range(n_features):
-            Xi = self._get_feature(X, feature_idx=i)
+            Xi = safe_indexing(X, i, axis=1)
             Xi = check_array(Xi, ensure_2d=False, dtype=None,
                              force_all_finite=needs_validation)
             X_columns.append(Xi)
 
-        return X_columns, n_samples, n_features
+        return X_columns
 
-    def _get_feature(self, X, feature_idx):
-        if hasattr(X, 'iloc'):
-            # pandas dataframes
-            return X.iloc[:, feature_idx]
-        # numpy arrays, sparse arrays
-        return X[:, feature_idx]
+    def _check_categories(self, n_features):
+        """Check categories are consistent with n_features"""
+        categories = self.categories
 
-    def _fit(self, X, handle_unknown='error'):
-        X_list, n_samples, n_features = self._check_X(X)
+        if (self.categories != 'auto' and
+                len(self.categories) != n_features):
+            raise ValueError("Shape mismatch: if categories is an array, "
+                             "it has to be of shape (n_features,).")
 
-        if self.categories != 'auto':
-            if len(self.categories) != n_features:
-                raise ValueError("Shape mismatch: if categories is an array,"
-                                 " it has to be of shape (n_features,).")
+        if isinstance(self.categories, str) and self.categories == 'auto':
+            categories = ['auto'] * n_features
+        else:
+            categories = self.categories
 
-        self.categories_ = []
+        return categories
 
-        for i in range(n_features):
-            Xi = X_list[i]
-            if self.categories == 'auto':
-                cats = _encode(Xi)
-            else:
-                cats = np.array(self.categories[i], dtype=Xi.dtype)
-                if Xi.dtype != object:
-                    if not np.all(np.sort(cats) == cats):
-                        raise ValueError("Unsorted categories are not "
-                                         "supported for numerical categories")
-                if handle_unknown == 'error':
-                    diff = _encode_check_unknown(Xi, cats)
-                    if diff:
-                        msg = ("Found unknown categories {0} in column {1}"
-                               " during fit".format(diff, i))
-                        raise ValueError(msg)
-            self.categories_.append(cats)
-
-    def _transform(self, X, handle_unknown='error'):
-        X_list, n_samples, n_features = self._check_X(X)
-
-        X_int = np.zeros((n_samples, n_features), dtype=np.int)
-        X_mask = np.ones((n_samples, n_features), dtype=np.bool)
+    def _fit_list(self, X_list, encoders):
+        """Fit encoders on X_list"""
+        assert len(X_list) == len(encoders)
```

Comment:
I would add an additional message in case we raise the AssertionError. It will not be really friendly.